### PR TITLE
OpenId Connect authentication for new management API

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description

This PR is the foundation for bringing in OpenId Connect as authentication for the new back-office. It is by no means feature complete, but it is fully functional as an MVP implementation. All known remaining tasks can be found in the backlog under _Management API - Authentication To-do_, including tasks for all TODOs in the code for this PR.

Since we're building the new back-office alongside the current, this new authentication scheme can co-exist with the current cookie based one for a foreseeable future. 

We have chosen [OpenIddict](https://github.com/openiddict/openiddict-core) as the framework for building our OpenId Connect server. For authorization we'll be using the _Authorization Code Flow with Proof Key for Code Exchange (PKCE)_, which is the recommended flow for modern single page applications. It looks something like this:

image

As per the OpenId Connect standard (and as illustrated above), the server will host its own authentication/login screen seperate from the client. For the time being we'll reuse the existing login dialog with a small tweak (see _login.controller.js_), which means we won't be blocked by having to create a new client application for login right away - we can create that down the road, thus also properly plan for handling 2FA and 3rd party login providers from the get-go.

### Testing this PR

_Remember to build the client before testing!_

#### Umbraco back-office

First and foremost make sure you can still login to the back-office as per usual.

#### Swagger UI

1. Protect an endpoint from the new management API - e.g. add `[Authorize(Policy = $"New{AuthorizationPolicies.SectionAccessForMediaTree}")]` to `MediaTreeControllerBase`.
    - _Note: $"New{AuthorizationPolicies.SectionAccessForMediaTree}" is one of the new policies that are created for OpenIddict as a proof-of-concept for testing. All policies will have to be reworked as part of finishing up the entire feature._
2. Verify that you can authorize against the endpoint using Swagger UI:
auth-swagger-ui

#### Postman

1. Configure a Postman request as follows to create new access tokens using the back-office authorization and token endpoints:
    - "Token Name": Whatever you fancy
    - "Grant Type": Authorization Code (With PKCE)
    - Tick "Authorize using browser"
    - "Auth URL": https://localhost:44331/umbraco/management/api/v1.0/security/back-office/authorize
    - "Access Token URL": https://localhost:44331/umbraco/management/api/v1.0/security/back-office/token
    - "Client ID": umbraco-postman
    - Leave "Client Secret" empty
    - "Code Challenge Method": SHA-256
    - "Code Verifier": Whatever you fancy
    - Leave "Scope" and "State" empty
    - "Client Authentication": Send client credentials in body
2. Hit "Get New Access Token" and verify that you can complete the authorization flow:
auth-postman
3. Verify that you can use the obtained access token to authenticate yourself against the protected endpoint from the Swagger test.

#### Simulated client login

As we do not yet have an actual client to test with, we will make do with simulating the flow in a few manual steps.

1. Make sure you're logged out of Umbraco (use an incognito browser if needed).
2. Call the authorize endpoint in a browser using these magic parameters: https://localhost:44331/umbraco/management/api/v1.0/security/back-office/authorize?client_id=umbraco-back-office&scope=offline_access&response_type=code&code_challenge=WZRHGrsBESr8wYFZ9sx0tPURuZgG2lmzyvWpwXPKz8U&code_challenge_method=S256
    - If the parameters look familiar, it's because they are basically the same as was configured in the Postman test above. In this case, the `code_challenge` is a hashed and encoded version of the code verifier "12345".
3. You will be redirected to the Umbraco login screen - log in.
4. You will be met by a "page not found" screen. The page URL will contain a `code` parameter - copy that.
auth redirect 404
5. Use Postman to obtain an access and a refresh token using the token endpoint
    - The token endpoint URL is (still): https://localhost:44331/umbraco/management/api/v1.0/security/back-office/token
    - The request must be a POST using "x-www-form-urlencoded" parameters:
        - "grant_type": authorization_code
        - "client_id": umbraco-back-office
        - "code": The code parameter from the page URL above
        - "code_verifier": 12345
auth token request
6. Verify that you can use the obtained access token to authenticate yourself against the protected endpoint from the Swagger test.
7. Verify that you can obtain a new access token using the refresh token obtained from the token endpoint.
    - You can reuse the Postman request from above, as it still a POST request to the same token endpoint. 
    - The parameters should be:
        - "grant_type": refresh_token
        - "client_id": umbraco-back-office
        - "refresh_token": The value of the obtained refresh token
8. Verify that you can use the new access token to authenticate yourself against the protected endpoint. 
    - At this point both access tokens will actually work, as old access tokens are not revoked when obtaining new ones. This is by design.
auth token refresh

### CODE REVIEW

File: src/Umbraco.Cms.ManagementApi/Security/BackOfficeApplicationManager.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        _applicationManager = applicationManager;
+        _webHostEnvironment = webHostEnvironment;
+        _clientSecretManager = clientSecretManager;
+        _backOfficeHost = securitySettingsMonitor.CurrentValue.BackOfficeHost;
+    }
+
+    public async Task EnsureBackOfficeApplicationAsync(Uri backOfficeUrl, CancellationToken cancellationToken = default)
+    {
+        if (backOfficeUrl.IsAbsoluteUri == false)
```
Comment (Zeegaan) [inline src/Umbraco.Cms.ManagementApi/Security/BackOfficeApplicationManager.cs:32]:
```suggestion
        if (backOfficeUrl.IsAbsoluteUri is false)
```

File: src/Umbraco.Cms.ManagementApi/Security/BackOfficeApplicationManager.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                cancellationToken);
+        }
+    }
+
+    private async Task CreateOrUpdate(OpenIddictApplicationDescriptor clientDescriptor, CancellationToken cancellationToken)
+    {
+        var identifier = clientDescriptor.ClientId ??
+                         throw new ApplicationException($"ClientId is missing for application: {clientDescriptor.DisplayName ?? "(no name)"}");
+        var client = await _applicationManager.FindByClientIdAsync(identifier, cancellationToken);
+        if (client == null)
```
Comment (Zeegaan) [inline src/Umbraco.Cms.ManagementApi/Security/BackOfficeApplicationManager.cs:116]:
```suggestion
        if (client is null)
```

File: src/Umbraco.Cms.ManagementApi/Security/BackOfficeApplicationManager.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        else
+        {
+            await _applicationManager.UpdateAsync(client, clientDescriptor, cancellationToken);
+        }
+    }
+
+    private async Task Delete(string identifier, CancellationToken cancellationToken)
+    {
+        var client = await _applicationManager.FindByClientIdAsync(identifier, cancellationToken);
+        if (client == null)
```
Comment (Zeegaan) [inline src/Umbraco.Cms.ManagementApi/Security/BackOfficeApplicationManager.cs:129]:
```suggestion
        if (client is null)
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Found some small nitpicky things, the rest LGTM 👍

**Comment by kjac** [comment]
@Zeegaan done done and done 😄

File: src/Umbraco.Core/Routing/UmbracoRequestPaths.cs [hunk 4]
```diff
@@ -51,6 +61,7 @@ public UmbracoRequestPaths(IOptions<GlobalSettings> globalSettings, IHostingEnvi
     ///         These are def back office:
     ///         /Umbraco/BackOffice     = back office
     ///         /Umbraco/Preview        = back office
+    ///         /Umbraco/Management/Api = back office
```
Comment (kjac) [inline src/Umbraco.Core/Routing/UmbracoRequestPaths.cs:64]:
```suggestion
```

File: src/Umbraco.New.Cms.Infrastructure/Umbraco.New.Cms.Infrastructure.csproj [hunk 5]
```diff
@@ -10,4 +10,8 @@
     <ProjectReference Include="..\Umbraco.New.Cms.Core\Umbraco.New.Cms.Core.csproj" />
     <ProjectReference Include="..\Umbraco.Infrastructure\Umbraco.Infrastructure.csproj" />
   </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="OpenIddict.Abstractions" Version="3.1.1" />
```
Comment (kjac) [inline src/Umbraco.New.Cms.Infrastructure/Umbraco.New.Cms.Infrastructure.csproj:15]: @kjac is this really necessary? Double check and clean up...

Comment (kjac) [reply src/Umbraco.New.Cms.Infrastructure/Umbraco.New.Cms.Infrastructure.csproj:15]: It is indeed necessary... the `OpenIddictCleanup` is located in this project, and it references the OpenIddict abstractions.

**Comment by Zeegaan** [comment]
Looks good and tests out great 🎉

**Pull request merged**